plugins/index-by-search/src/main/java/org/elasticsearch/plugin/indexbysearch/AbstractAsyncScrollAction.java

```diff
@@ -0,0 +1,339 @@
+        return updated.get();
+    }
+
+    /**
+     * Count of documents created.
+     */
+    public long created() {
+        return created.get();
+    }
+
+    /**
+     * Count of successful delete operations.
+     */
+    public long deleted() {
+        return deleted.get();
+    }
+
+    /**
+     * The number of bulk requests this request has completed.
+     */
+    public int batches() {
+        return batches.get();
+    }
+
+    public long versionConflicts() {
+        return versionConflicts.get();
+    }
+
+    public long successfullyProcessed() {
+        return updated.get() + created.get() + deleted.get();
+    }
+
+    public List<Failure> failures() {
+        return unmodifiableList(failures);
+    }
+
+    void initialSearch() {
+        try {
+            startTime.set(System.nanoTime());
+            if (logger.isDebugEnabled()) {
+                logger.debug("executing initial scroll against {}{}",
+                        firstSearchRequest.indices() == null ? "all indices" : firstSearchRequest.indices(),
+                        firstSearchRequest.types() == null || firstSearchRequest.types().length == 0 ? "" : firstSearchRequest.types());
+            }
+            searchAction.execute(firstSearchRequest, new ActionListener<SearchResponse>() {
+                @Override
+                public void onResponse(SearchResponse response) {
+                    logger.debug("[{}] documents match query", response.getHits().getTotalHits());
+                    total.set(response.getHits().getTotalHits());
+                    onScrollResponse(response);
+                }
+
+                @Override
+                public void onFailure(Throwable e) {
+                    logger.error("failed while executing the initial scroll request", e);
+                    listener.onFailure(e);
+                }
+            });
+        } catch (Throwable t) {
+            finishHim(t);
+        }
+    }
+
+    void onScrollResponse(final SearchResponse searchResponse) {
+        try {
+            scroll.set(searchResponse.getScrollId());
+            SearchHit[] docs = searchResponse.getHits().getHits();
+            logger.debug("scroll returned [{}] documents with a scroll id of [{}]", docs.length, searchResponse.getScrollId());
+            if (docs.length == 0) {
+                finishHim(null);
+                return;
+            }
+            List<SearchHit> docsIterable = Arrays.asList(docs);
+            if (mainRequest.size() != -1) {
+                // Truncate the docs if we have more than the request size
+                long remaining = max(0, mainRequest.size() - successfullyProcessed());
+                if (remaining <= docs.length) {
+                    if (remaining < docs.length) {
+                        docsIterable = docsIterable.subList(0, (int) remaining);
+                    }
+                }
+            }
+            BulkRequest request = buildBulk(docsIterable);
```

Comment:
I like the simplicity of this code but I'm wondering if we should use https://github.com/elastic/elasticsearch/pull/14829 since it will handle rejections.